Perf api check

kalash9747 opened the pull request:
Добавил новый модуль perf, подключил jooq, сделал тесты  добавления пользователя и покупки автомобиля пользователем без дома

![image](https://user-images.githubusercontent.com/82981820/172016881-6d9a1084-3621-4fb1-a916-57ccf7774c0c.png)
![image](https://user-images.githubusercontent.com/82981820/172016904-922c1fc7-19d6-496b-9474-dbcae431cb80.png)

kalash9747 committed:
Merge branch 'authCloudMailUi' into perfApiCheck

kalash9747 committed:
added new module perf
added JsonHelper
added JooqQuery
added PerfDbExecutor
added PerfApi
added addUserCheck
added buyCarWithoutHouseCheck

kalash9747 committed:
added jooq to Deps
update README
update TESTRUN

dKusakin commented on perf/build.gradle:
```diff
@@ -0,0 +1,52 @@
+}
+
+allure {
+    autoconfigure = true
```
Плагин аллюра решил старый использовать?

kalash9747 replied:
да, пока хотел на старых а переезд на новые версии аллюра и селенида  хотел сделать по завершении всего задания

dKusakin commented on perf/build.gradle:
```diff
@@ -0,0 +1,52 @@
+    testImplementation junitApi
+    testImplementation allureJunit
+    testImplementation allureAssertJ
+    testImplementation allureCommandLine
```
Зачем тебе commandline внутри проекта?

dKusakin commented on perf/build.gradle:
```diff
@@ -0,0 +1,52 @@
+            generationTool {
+                jdbc {
+                    driver = 'org.postgresql.Driver'
+                    url = 'jdbc:postgresql://77.50.236.203:4832/pflb_trainingcenter'
```
Хм, затягивать бы это из конфигов..
Не выйдет как-то из ресурсов подтянуть?

dKusakin commented on perf/src/main/java/encryption/PerfUserRole.java:
```diff
@@ -0,0 +1,6 @@
+package encryption;
+
+public enum PerfUserRole implements UserRole {
+    perfDBReader,
```
Вообще это константы, а константы принято в другом стиле записывать

dKusakin commented on perf/src/main/java/database/PerfDbExecutor.java:
```diff
@@ -0,0 +1,75 @@
+                .and(PERSON.SECOND_NAME.eq(person.getSecondName()))
+                .and(PERSON.AGE.eq(person.getAge()))
+                .and(PERSON.MONEY.eq(person.getMoney()).toString())
+                .and(PERSON.SEX.eq(true))
```
А чего это ты тут дам обделил?)

dKusakin commented on perf/src/test/java/usersTests/PerfApiTest.java:
```diff
@@ -0,0 +1,76 @@
+    private static final PerfDbExecutor dbExecutor = new PerfDbExecutor(getUser(perfDBReader));
+    private static final PerfApi perfApi = new PerfApi();
+
+    @DisplayName("Проверка добаления пользователя")
```
доба**в**ления

dKusakin commented on perf/src/test/java/usersTests/PerfApiTest.java:
```diff
@@ -0,0 +1,76 @@
+    @Owner("Калашников Владислав Александрович")
+    @Test
+    void addUserCheck() {
+        PersonRecord person = new PersonRecord(null, 25, "Tony", valueOf(247000), "Stark", true, null);
```
"Тест не должен зависеть от конкретных тестовых данных"
Посмотри в сторону [`Faker`'а](https://github.com/DiUS/java-faker) или чего-то подобного

dKusakin commented on perf/src/test/java/usersTests/PerfApiTest.java:
```diff
@@ -0,0 +1,76 @@
+        perfApi.addUser(new PersonRq(person))
+                .shouldBeStatusCode(201);
+        step("Проверить, что в базе появился новый пользователь", () ->
+                assertEquals(countRecordsBefore + 1, dbExecutor.getCountPersonRecords(person),
```
Хотелось бы не только количество, но и ответ проверить, там ведь прилетает созданный пользак с id.
Мб сначала максимальный id проверять, а потом проверять, что у вернувшегося он не меньше? Ну и потом соответствие данных ещё

dKusakin commented on perf/src/test/java/usersTests/PerfApiTest.java:
```diff
@@ -0,0 +1,76 @@
+     * Возвращает самый дешевый автомобиль
+     */
+    private CarRecord getCheapestCar() {
+        return dbExecutor.getCarRecords().stream()
```
А я правильно понял, что сначала нашли все автомобили запросом, а потом фильтранули результат уже в джаве? Мб стоило сразу в БД нужный запрос кидать?

kalash9747 replied:
Да правильно, здесь я просто хотел некоторые кейсы стримами , а некоторые запросом в БД, так как третий кейс подразумевает более сложный запрос , его решил запросом в БД ,а эти стримами,  если это принципиально могу переделать если нет, то оставил бы)

dKusakin commented on perf/src/test/java/usersTests/PerfApiTest.java:
```diff
@@ -0,0 +1,76 @@
+     * Возвращает самого богатого пользователя
+     */
+    private PersonRecord getRichestPerson() {
+        return dbExecutor.getPersonRecords().stream()
```
Тут вопрос аналогичный автомобилям

kalash9747 replied:
оставил самого богатого пользователя к самому дешевому авто

dKusakin commented on perf/src/test/java/usersTests/PerfApiTest.java:
```diff
@@ -0,0 +1,76 @@
+    void buyCarWithoutHouseCheck() {
+        PersonRecord person = getRichestPerson();
+        CarRecord car = getCheapestCar();
+        assumeTrue(person.getMoney().compareTo(car.getPrice()) >= 0,
```
Выход интересный, просто взять самого богатого человека и самый дешёвый автомобиль)
Я бы исхитрялся через запрос, т.е. декартово произведение человеков и автомобилей, а потом их фильтрация по "автомобиль не должен принадлежать человеку и кол-во денег человека >= стоимость автомобиля" и выбор любой полученной записи

kalash9747 replied:
оставил самого богатого пользователя к самому дешевому авто

dKusakin commented on perf/src/test/java/usersTests/PerfApiTest.java:
```diff
@@ -0,0 +1,76 @@
+        PersonRecord person = getRichestPerson();
+        CarRecord car = getCheapestCar();
+        assumeTrue(person.getMoney().compareTo(car.getPrice()) >= 0,
+                "В базе данных нет пользователя без дома, у которого хватает денег, хотя бы на самую дешевую машину");
```
Между "денег" и "хотя бы" запятая не нужна

dKusakin commented on util/src/main/java/api/BaseHttpMethods.java:
```diff
@@ -81,4 +84,18 @@ public HttpResponseFacade post(URI uri, BodyPublisher body, String... headers) {
+    /**
+     * Отправить POST запрос без тела и заголовков
+     */
+    public HttpResponseFacade post(URI uri) {
```
Что-то сразу не подумал, но если в 70 и 78 строке написать не просто `headers == null`, а расширить до `headers == null || headers.length < 1`, то мб от части методов удастся избавиться, ведь `vararg` поддерживает и отсутствие аргументов

kalash9747 committed:
added javaFaker
rename Roles
rename PersonRq
added getMaxPersonId
added filter PersonId
added checkNewPerson

kalash9747 committed:
fix filter PersonId

dKusakin commented on perf/src/main/java/models/PersonInfo.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    public double getMoney() {
+        return money.doubleValue();
```
Сразу не обратил внимание. Зачем перегонять в `double`?

dKusakin commented on perf/src/main/java/models/PersonInfo.java:
```diff
@@ -0,0 +1,105 @@
+        return money.doubleValue();
+    }
+
+    public PersonInfo setMoney(double money) {
```
Аналогично, почему сразу не из `BigDecimal`?

dKusakin commented on perf/src/test/java/usersTests/PerfApiTest.java:
```diff
@@ -0,0 +1,107 @@
+        Faker faker = new Faker();
+        long maxPersonIdBefore = dbExecutor.getMaxPersonId();
+
+        PersonInfo personRq = new PersonInfo(new PersonRecord(null, faker.number().numberBetween(18, 100),
```
Мб лучше через цепочку сеттеров? А то эта штука как-то тяжко читается с кучей параметров

kalash9747 committed:
change type money
